File: linkList/100-19-removeNthFromEnd.cpp
/**
 * Definition for singly-linked list.*/
  struct ListNode {
      int val;
      ListNode *next;
      ListNode() : val(0), next(nullptr) {}
      ListNode(int x) : val(x), next(nullptr) {}
      ListNode(int x, ListNode *next) : val(x), next(next) {}
  };
  
 /* Given the head of a linked list, remove the nth node from the end of the list and return its head.
 */
#include<iostream>
// #include<>
using std::cout;


class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *slow=head, *fast=head;
        int cur=1;
        while(cur<=n)
        {
            if(fast->next)
                fast=fast->next;
            else
                return head->next;
            
            cur+=1;
        }
        while(fast->next)
        {
            slow=slow->next;
            fast=fast->next;
        }
        slow->next=slow->next->next;
        return head;
    }
};


int main()
{
    ListNode *first=new ListNode(1);
    
    ListNode *second=new ListNode(2);
    ListNode *third=new ListNode(3);
    ListNode *four=new ListNode(4);
    third->next=four;
    second->next=third;
    first->next=second;
    ListNode *res=Solution().removeNthFromEnd(first,2);
    while(res)
    {
        cout<<res->val;
        res=res->next;
    }
}

File: linkList/100-19-removeNthFromEnd.py
from ListNode import ListNode
'''
我们可以设想假设设定了双指针 p 和 q 的话，当 q 指向末尾的 NULL，p 与 q 之间相隔的元素个数为 n 时，那么删除掉 p 的下一个指针就完成了要求。

设置虚拟节点 dummyHead 指向 head
设定双指针 p 和 q，初始都指向虚拟节点 dummyHead
移动 q，直到 p 与 q 之间相隔的元素个数为 n
同时移动 p 与 q，直到 q 指向的为 NULL
将 p 的下一个节点指向下下个节点

'''
def removeNthFromEnd(header,n):
    i=1
    p,q=header,header
    while(i<=n):
        i+=1
        if not p.next:
            return p.next
        else:
            p=p.next
    while p.next:
        p=p.next
        q=q.next
    q.next=q.next.next
    return header

first=ListNode(1)
second=ListNode(2)
third=ListNode(3)
four=ListNode(4)
five=ListNode(5)
first.setNext(second)
second.setNext(third)
third.setNext(four)
four.setNext(five)

res=removeNthFromEnd(first,2)
while(res):
    print(res.val,"->")
    res=res.next
